File: backend/app/api/v1/endpoints/ai.py
from typing import Any
from fastapi import APIRouter, HTTPException, Depends
from app.api.deps import get_current_user
from app.schemas.ai import (
    GigGenerateRequest, GigGenerateResponse,
    GigEditRequest, GigGenerateResponse as GigEditResponse,
    PitchGenerateRequest, PitchGenerateResponse,
    ReviewGenerateRequest, ReviewGenerateResponse,
    BioGenerateRequest, BioGenerateResponse,
    HelperChatRequest, HelperChatResponse,
)
from app.services.ai_service import ai_service

router = APIRouter()


@router.post("/generate-gig", response_model=GigGenerateResponse)
def generate_gig(request: GigGenerateRequest, current_user: Any = Depends(get_current_user)):
    """Generate a full gig post from a short description using AI."""
    try:
        result = ai_service["generate_gig"](request.prompt)
        return GigGenerateResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"AI generation failed: {str(e)}")


@router.post("/edit-gig", response_model=GigGenerateResponse)
def edit_gig(request: GigEditRequest, current_user: Any = Depends(get_current_user)):
    """Edit an existing gig post with natural language instructions."""
    try:
        current_gig = {
            "title": request.title,
            "description": request.description,
            "tags": request.tags,
            "category": request.category,
        }
        result = ai_service["edit_gig"](current_gig, request.instruction)
        return GigGenerateResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"AI edit failed: {str(e)}")


@router.post("/generate-pitch", response_model=PitchGenerateResponse)
def generate_pitch(request: PitchGenerateRequest, current_user: Any = Depends(get_current_user)):
    """Generate a personalized application pitch using AI."""
    try:
        pitch = ai_service["generate_pitch"](
            request.job_title,
            request.job_description,
            request.user_skills,
            request.user_bio,
        )
        return PitchGenerateResponse(pitch=pitch)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"AI pitch generation failed: {str(e)}")


@router.post("/generate-review", response_model=ReviewGenerateResponse)
def generate_review(request: ReviewGenerateRequest, current_user: Any = Depends(get_current_user)):
    """Generate a review comment based on gig context and rating."""
    if not 1 <= request.rating <= 5:
        raise HTTPException(status_code=400, detail="Rating must be between 1 and 5")
    try:
        comment = ai_service["generate_review"](request.job_title, request.rating)
        return ReviewGenerateResponse(comment=comment)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"AI review generation failed: {str(e)}")


@router.post("/generate-bio", response_model=BioGenerateResponse)
def generate_bio(request: BioGenerateRequest, current_user: Any = Depends(get_current_user)):
    """Generate a professional bio from skills."""
    try:
        bio = ai_service["generate_bio"](request.skills, request.username)
        return BioGenerateResponse(bio=bio)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"AI bio generation failed: {str(e)}")


@router.post("/chat", response_model=HelperChatResponse)
def helper_chat(request: HelperChatRequest, current_user: Any = Depends(get_current_user)):
    """AI helper agent conversation."""
    try:
        messages = [{"role": m.role, "content": m.content} for m in request.messages]
        result = ai_service["helper_chat"](messages)
        return HelperChatResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"AI chat failed: {str(e)}")

File: frontend/src/api/ai.ts
import { api } from './client'

// --- Types ---
export interface GigGenerateResponse {
    title: string
    description: string
    tags: string[]
    category: string
}

export interface PitchGenerateResponse {
    pitch: string
}

export interface ReviewGenerateResponse {
    comment: string
}

export interface BioGenerateResponse {
    bio: string
}

export interface ChatMessage {
    role: 'user' | 'assistant'
    content: string
}

export interface HelperChatResponse {
    reply: string
    escalate: boolean
    email: string | null
}

// --- API Functions ---
export const aiApi = {
    generateGig: async (prompt: string): Promise<GigGenerateResponse> => {
        const response = await api.post<GigGenerateResponse>('/ai/generate-gig', { prompt })
        return response.data
    },

    editGig: async (title: string, description: string, tags: string[], category: string, instruction: string): Promise<GigGenerateResponse> => {
        const response = await api.post<GigGenerateResponse>('/ai/edit-gig', {
            title, description, tags, category, instruction
        })
        return response.data
    },

    generatePitch: async (jobTitle: string, jobDescription: string, userSkills: string[], userBio: string): Promise<PitchGenerateResponse> => {
        const response = await api.post<PitchGenerateResponse>('/ai/generate-pitch', {
            job_title: jobTitle,
            job_description: jobDescription,
            user_skills: userSkills,
            user_bio: userBio,
        })
        return response.data
    },

    generateReview: async (jobTitle: string, rating: number): Promise<ReviewGenerateResponse> => {
        const response = await api.post<ReviewGenerateResponse>('/ai/generate-review', {
            job_title: jobTitle,
            rating,
        })
        return response.data
    },

    generateBio: async (skills: string[], username: string = ''): Promise<BioGenerateResponse> => {
        const response = await api.post<BioGenerateResponse>('/ai/generate-bio', {
            skills,
            username,
        })
        return response.data
    },

    helperChat: async (messages: ChatMessage[]): Promise<HelperChatResponse> => {
        const response = await api.post<HelperChatResponse>('/ai/chat', { messages })
        return response.data
    },
}
